rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+        let mut err = ErrorMsg::new(lib.borrow());
+        let rng = unsafe {
+            lib.borrow()
+                .0
+                .bs_rng_construct(seed as c_uint, err.as_ptr())
+        };
+        if let Some(rng) = NonNull::new(rng) {
+            drop(err);
+            Ok(Self { rng, lib })
+        } else {
+            Err(BridgeStanError::ConstructFailed(err.message()))
+        }
+    }
+}
+
+struct ErrorMsg<'lib> {
+    msg: *mut c_char,
+    lib: &'lib StanLibrary,
+}
+
+impl<'lib> Drop for ErrorMsg<'lib> {
+    fn drop(&mut self) {
+        if !self.msg.is_null() {
+            unsafe { self.lib.0.bs_free_error_msg(self.msg) };
+        }
+    }
+}
+
+impl<'lib> ErrorMsg<'lib> {
+    fn new(lib: &'lib StanLibrary) -> Self {
+        Self {
+            msg: std::ptr::null_mut(),
+            lib,
+        }
+    }
+
+    fn as_ptr(&mut self) -> *mut *mut c_char {
+        &mut self.msg
+    }
+
+    /// Return the error message as a String.
+    ///
+    /// Panics if there was no error message.
+    fn message(&self) -> String {
+        NonNull::new(self.msg)
+            .map(|msg| {
+                unsafe { CStr::from_ptr(msg.as_ptr()) }
+                    .to_string_lossy()
+                    .to_string()
+            })
+            .expect("Stan returned an error but no error message")
+    }
+}
+
+impl<T: Borrow<StanLibrary>> Model<T> {
+    /// Create a new instance of the compiled Stan model.
+    /// Data is specified as a JSON file at the given path, or empty for no data
+    /// Seed and chain ID are used for reproducibility.
+    pub fn new<D: AsRef<CStr>>(lib: T, data: Option<D>, seed: u32) -> Result<Self> {
+        let mut err = ErrorMsg::new(lib.borrow());
+
+        let data_ptr = data
+            .as_ref()
+            .map(|data| data.as_ref().as_ptr())
+            .unwrap_or(null());
+        let model = unsafe {
+            lib.borrow()
+                .0
+                .bs_model_construct(data_ptr, seed, err.as_ptr())
+        };
+        // Make sure data lives until here
+        drop(data);
+
+        if let Some(model) = NonNull::new(model) {
+            drop(err);
+            let model = Self { model, lib };
+            // If STAN_THREADS is not true, the safty guaranties we are
+            // making would be incorrect
+            let info = model.info();
+            if !info.to_string_lossy().contains("STAN_THREADS=true") {
+                Err(BridgeStanError::StanThreads(
+                    info.to_string_lossy().into_owned(),
+                ))
+            } else {
+                Ok(model)
+            }
+        } else {
+            Err(BridgeStanError::ConstructFailed(err.message()))
+        }
+    }
+
+    /// Return a reference to the underlying stan library
+    pub fn ref_library(&self) -> &StanLibrary {
+        self.lib.borrow()
+    }
+
+    pub fn new_rng(&self, seed: u32) -> Result<Rng<&StanLibrary>> {
+        Rng::new(self.ref_library(), seed)
+    }
```

Comment:
We provide this helper function in the other languages primarily because they don't have first-class handling of a StanLibrary-like object. So this probably unnecessary here, though it does have the nice property of encouraging code which would ensure that RNGs come from the same underlying library as the model.

(Aside: Is that [possible to have?] enforced by the type system? I don't know what happens if you feed an RNG from `bernoulli_model.so` to `jacobian_model.so`, but I imagine it might get weird!)

Reply:
I added this when the seed was model specific, and didn't think about it after that changed.

I think it's not bad to have, but not necessary.

Good question about using cross-lib rng objects. I think I would like to prevent that, but I can't think of a type system check for it. I guess we can store an id in the library and the rng object and do a runtime check?